### PR TITLE
ch4-complexity-introduction: done

### PR DESCRIPTION
**Pull request by ghbooth12**

### CODE REVIEW

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 1]
```diff
@@ -0,0 +1,87 @@
+1. Describe an analogy for relating an algorithm that has efficiency O(1) and another algorithm that has O(2^n). An example would be:
+  An algorithm that has O(1) is a Cheetah and an algorithm that has O(2^n) is a Snail.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:2]: I love this analogy! 👍

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    An algorithm that has O(1) takes constant time to execute no matter how big the input size is. An algorithm that has O(2^n) takes exponential time to execute even with a small size of input. Thus the former algorithm is like the latest model of CPU and the latter algorithm is like the first computer 'ENIAC'.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:4]: Yes!

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+2. In plain English, what is the best case scenario for binary search?
+
+    The target item is located in the middle of the collection.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:10]: Can you elaborate here?

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+3. In plain English, what is the worst-case scenario for binary search?
+
+    The target item is located in the beginning of the collection or at the end of the collection.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:16]: Why?

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+
+4. In plain English, what is the bounded-case scenario for binary search?
+
+    The target item is located between the middle item and the leftmost/rightmost item.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:21]: Please elaborate.

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    6,64
+    7,128
+    8,256
+    9,512
+    10,1024
+
+    <functional equation>
+    f(n) = 2^n
+
+    => file name: graph_exponential_eq.png
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:41]: Nifty.

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+6. What's the asymptotic limit as n approaches infinity for the function defined by the values above?
+
+    The limit of 2 to the n-th power as n approaches infinity is infinity.
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:47]: Yep!

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+7. What is the Big-O of an algorithm that has the data points above?
+
+    O(2^n)
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:53]: 👍

File: 04-complexity-introduction/intro_complexity_answers.txt (hunk 3 above)
Comment (ghbooth12) [reply 04-complexity-introduction/intro_complexity_answers.txt:10]:
``` ruby
def binary_search(array, target)
  low = 0
  high = array.length - 1

  while low <= high
    mid = (low + high) / 2
    if array[mid] > target
      high = mid - 1
    elsif array[mid] < target
      low = mid + 1
    else
      return mid
    end
  end
end
```

According to the code above, if the 'target' is at the 'mid' index the while loop ends and returns the mid. In other words, the best case scenario is that the first element the method looks at is the one it is looking for. In Binary Search, it always picks the element that is in the middle of the collection for comparing with the target item. So the best case scenario in binary search is that the target element is in the middle of collection.

File: 04-complexity-introduction/intro_complexity_answers.txt (hunk 4 above)
Comment (ghbooth12) [reply 04-complexity-introduction/intro_complexity_answers.txt:16]: Binary Search picks the element that is in the middle of the collection for checking if it is equal to the target element. If the target element is at the leftmost index or the rightmost index, that would be the worst case since the 'middle' index can't be at the leftmost index or the rightmost index (for the large size of collection).

File: 04-complexity-introduction/intro_complexity_answers.txt (hunk 5 above)
Comment (ghbooth12) [reply 04-complexity-introduction/intro_complexity_answers.txt:21]: The bounded-case scenario is the performance between the best-case and the worst-case. So when an element that the method is looking at isn't at the middle of collection nor at the leftmost-rightmost index, that is the bounded-case.

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+9. Create a graph from the output using Google Sheets or other graphing software. Analyze the graph and denote its Big-O somewhere on the graph.
+
+    => file name: graph_linear_search_complexity.png
+
+
+
+10. What is the Big-O of binary search?
+12. What is the Big-Ө of binary search?
+11. What is the Big-Ω of binary search?
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:71]: Correct!

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+9. Create a graph from the output using Google Sheets or other graphing software. Analyze the graph and denote its Big-O somewhere on the graph.
+
+    => file name: graph_linear_search_complexity.png
+
+
+
+10. What is the Big-O of binary search?
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:69]: Correct!

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+12. What is the Big-Ө of binary search?
```
Comment (robmadden) [inline 04-complexity-introduction/intro_complexity_answers.txt:70]: Correct!

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+11. What is the Big-Ω of binary search?
+
+    <worst-case>
+      f(n) = lg (n) + 1    # lg: log base-2
```
Comment (ghbooth12) [inline 04-complexity-introduction/intro_complexity_answers.txt:74]: O(lg(n))

This represents the complexity of the worst-case scenario for your algorithm.

File: 04-complexity-introduction/intro_complexity_answers.txt [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+      O(f(n)) = O(lg(n) + 1) = O(lg n) (describes upper bound)
+      Ω(f(n)) = Ω(lg(n) + 1) = Ω(lg n) (descirbes lower bound)
+
+      Thus binary search for the worst-case is Ө(lg n). (describes tight bound)
+
+    <best-case>
+      f(n) = 1    # first element is the one it is looking for
```
Comment (ghbooth12) [inline 04-complexity-introduction/intro_complexity_answers.txt:82]: Ω(1)

This represents the best case performance of your algorithm.
